File: Exercism Codes/PHP/Connect.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

function winner(array $lines): ?string
{
    $rows = array_map(function($x) {
        return explode(" ", trim($x));
    }, $lines);

    $check = function($row, $col, $visited, $l) use ($rows, &$check) {
        if ($l == "X" ? $col == count($rows[0]) : $row == count($rows)) {
            return true;
        }
        if (!($row >= 0 && $row < count($rows)) || !($col >= 0 && $col < count($rows[0])) || $visited[$row][$col]) {
            return false;
        }
        $visited[$row][$col] = true;

        if ($rows[$row][$col] == $l) {
            return $check($row + 1, $col, $visited, $l) ||
                $check($row - 1, $col, $visited, $l) ||
                $check($row, $col + 1, $visited, $l) ||
                $check($row, $col - 1, $visited, $l) ||
                $check($row + 1, $col - 1, $visited, $l) ||
                $check($row - 1, $col + 1, $visited, $l);
        }

        return false;
    };

    $xWins = false;
    $oWins = false;

    for ($x = 0; $x < count($rows); $x++) {
        if ($rows[$x][0] == "X") {
            $wins = $check($x, 0, createVisited($rows), "X");
            if ($wins) {
                $xWins = true;
                break;
            }
        }
    }

    for ($o = 0; $o < count($rows[0]); $o++) {
        if ($rows[0][$o] == "O") {
            $wins = $check(0, $o, createVisited($rows), "O");
            if ($wins) {
                $oWins = true;
                break;
            }
        }
    }

    if ($xWins) {
        return "black";
    }
    else if ($oWins) {
        return "white";
    }
    else {
        return null;
    }
}

function createVisited(array $rows): array {
    $result = [];
    
    for ($x = 0; $x < count($rows); $x++) {
        $z = [];
        for ($y = 0; $y < count($rows[$x]); $y++) {
            $z[$y] = false;
        }
        $result[$x] = $z;
    }

    return $result;
}

File: Exercism Codes/C#/Connect.cs
public enum ConnectWinner
{
    White,
    Black,
    None
}

public class Connect
{
    private ConnectWinner winner;

    public Connect(string[] input)
    {
        List<List<string>> rows = input.Select(line => line.Trim().Split(' ').ToList()).ToList();

        bool Check(int row, int col, List<List<bool>> visited, char l)
        {
            if (l == 'X' ? col == rows[0].Count : row == rows.Count)
            {
                return true;
            }

            if (!(row >= 0 && row < rows.Count) || !(col >= 0 && col < rows[0].Count) || visited[row][col])
            {
                return false;
            }

            visited[row][col] = true;

            if (rows[row][col][0] == l)
            {
                return Check(row + 1, col, visited, l) ||
                        Check(row - 1, col, visited, l) ||
                        Check(row, col + 1, visited, l) ||
                        Check(row, col - 1, visited, l) ||
                        Check(row - 1, col + 1, visited, l) ||
                        Check(row + 1, col - 1, visited, l);
            }

            return false;
        }

        bool xWins = false;
        bool oWins = false;

        for (int x = 0; x < rows.Count; x++)
        {
            if (rows[x][0] == "X")
            {
                List<List<bool>> visited = rows.Select(r => r.Select(_ => false).ToList()).ToList();
                if (Check(x, 0, visited, 'X'))
                {
                    xWins = true;
                    break;
                }
            }
        }

        for (int o = 0; o < rows[0].Count; o++)
        {
            if (rows[0][o] == "O")
            {
                List<List<bool>> visited = rows.Select(r => r.Select(_ => false).ToList()).ToList();
                if (Check(0, o, visited, 'O'))
                {
                    oWins = true;
                    break;
                }
            }
        }

        if (oWins)
        {
            winner = ConnectWinner.White;
        }
        else
        {
            winner = xWins ? ConnectWinner.Black : ConnectWinner.None;
        }
    }

    public ConnectWinner Result() => winner;
}